src/main/java/com/google/api/gax/retrying/RetryingFutureImpl.java

```diff
@@ -0,0 +1,183 @@
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package com.google.api.gax.retrying;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.api.gax.core.ApiFutureCallback;
+import com.google.common.util.concurrent.AbstractFuture;
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import java.util.concurrent.Callable;
+import java.util.concurrent.Future;
+
+/**
+ * For internal use only.
+ *
+ * <p>
+ * This class is the key component of the retry logic. It implements the {@link RetryingFuture}
+ * facade interface, and does the following:
+ *
+ * <ul>
+ * <li>Schedules the next attempt in case of a failure using the callback chaining technique.
+ * <li>Terminates retrying process if no more retries are accepted.
+ * <li>Propagates future cancellation in both directions (from this to the attempt and from the
+ * attempt to this)
+ * </ul>
+ *
+ * This class is thread-safe.
+ */
+class RetryingFutureImpl<ResponseT> extends AbstractFuture<ResponseT>
+    implements RetryingFuture<ResponseT> {
+
+  private final Object lock = new Object();
+  private final Callable<ResponseT> callable;
+
+  private final RetryAlgorithm retryAlgorithm;
+  private final RetryingExecutor<ResponseT> retryingExecutor;
+
+  private volatile TimedAttemptSettings attemptSettings;
+  private volatile AttemptFutureCallback attemptFutureCallback;
+
+  RetryingFutureImpl(
+      Callable<ResponseT> callable,
+      RetryAlgorithm retryAlgorithm,
+      RetryingExecutor<ResponseT> retryingExecutor) {
+    this.callable = checkNotNull(callable);
+    this.retryAlgorithm = checkNotNull(retryAlgorithm);
+    this.retryingExecutor = checkNotNull(retryingExecutor);
+
+    this.attemptSettings = retryAlgorithm.createFirstAttempt();
+  }
+
+  @Override
+  public boolean cancel(boolean mayInterruptIfRunning) {
+    synchronized (lock) {
+      if (attemptFutureCallback != null) {
+        if (attemptFutureCallback.attemptFuture.cancel(mayInterruptIfRunning)) {
+          super.cancel(mayInterruptIfRunning);
+        }
+        return isCancelled();
+      } else {
+        return super.cancel(mayInterruptIfRunning);
+      }
+    }
+  }
+
+  @Override
+  public void setAttemptFuture(Future<ResponseT> attemptFuture) {
+    if (isDone()) {
+      return;
+    }
+    synchronized (lock) {
+      if (isDone()) {
+        return;
+      }
+      if (attemptFuture != null) {
+        attemptFutureCallback = new AttemptFutureCallback(attemptFuture);
+        Futures.addCallback((ListenableFuture) attemptFuture, attemptFutureCallback);
```

Comment:
Yes, it requires that future to be Guava's. It is also a package-private class and it is mostly an implementation detail. 

During its life cycle, RetryingFuture wraps 2 different types of future for each type (one after another): first it wraps the ScheduledExecutor's FutureTask, when the task is waiting for execution in some time in future, after that the RetryingFuture replaces it with ApiFuture returned by the gax.grpc's FutureCallable. One of the reasons, why I prefer keeping generic Future interface is to indicate the fact that they can be really different futures.

Main reason why I don't want more layers is my experience with guava's futures so far. They work well for 99% cases, but when it comes to corner cases they are not that good. Specifically guava usually wraps classes from `java.util.concurrent` but does it with limitations, which are essential for cases like retries. Specifically check this: https://github.com/google/guava/blob/master/guava/src/com/google/common/util/concurrent/MoreExecutors.java#L559
i.e. guava does not cancel its wrapper future, when the wrapped delegate is canceled. Case like this makes me want to avoid extra wrapping when possible.

I've changed it to ApiFuture. But I still don't like it, because now it is harder to trace the calls and make some correctness guarantees.